lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,416 @@
+                        logger(logging.ERROR, '(%s) %s', rse['rse'], str(error))
+                    else:
+                        updated = True
+
+            reap_pending = False
+
+        # 2. Condition on the lock count
+        if replica['lock_cnt'] != 0:
+            # 2.1 No actual lock -> Reset the lock count
+            try:
+                locks = get_replica_locks(replica['scope'], replica['name'],
+                                          restrict_rses=[rse['id']])
+            except NoResultFound:
+                logger(logging.WARNING,
+                       '(%s) Replica %s:%s disappeared during cleanup',
+                       rse['rse'], replica['scope'], replica['name'])
+            else:
+                if len(locks) == 0:
+                    logger(logging.WARNING,
+                           '(%s) Replica %s:%s has lock count %s but zero actual locks. Please'
+                           ' fix the counts.',
+                           rse['rse'], replica['scope'], replica['name'], replica['lock_cnt'])
+
+            reap_pending = False
+
+        # 3. Tombstone missing or in the future -> RIP now
+        if replica['tombstone'] is None or replica['tombstone'] >= datetime.utcnow():
+            logger(logging.INFO, '(%s) Marking tombstone of replica %s:%s as UTCNOW.',
+                   rse['rse'], replica['scope'], replica['name'])
+
+            try:
+                set_tombstone(rse['id'], replica['scope'], replica['name'],
+                              tombstone=datetime.utcnow())
+            except ReplicaNotFound as error:
+                logger(logging.WARNING, '(%s) %s', rse['rse'], str(error))
+            else:
+                updated = True
+
+            reap_pending = False
+
+        if updated:
+            num_updated += 1
+
+        if reap_pending:
+            num_waiting_reap += 1
+
+    logger(logging.INFO, '(%s) %s replicas have been updated. %s replicas are pending deletion.'
+           ' %s replicas remain unhandled.',
+           rse['rse'], num_updated, num_waiting_reap,
+           num_processed - num_updated - num_waiting_reap)
+
+
+# Condition functions
+
+def _is_being_deleted(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check if the rule is already set to be deleted."""
+    if rule['expires_at'] is not None and rule['expires_at'] < datetime.utcnow():
+        logger(logging.DEBUG,
+               '(%s) Rule %s for %s:%s is bound for deletion',
+               rse['rse'], rule['id'], rule['scope'], rule['name'])
+        return True
+    return False
+
+
+def _has_trivial_rse_expression(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check for a trivial rse_expression."""
+    return rule['rse_expression'] == rse['rse']
+
+
+def _has_last_replica_on_rse(
```

Comment:
Can you suggest a better function name?

Reply:
has_only_one_replica_on_rse ?? I'm out of ideas..

Reply:
Reducing the function to its base components:

```python
def f(rule, rse):
    locks = get_replica_locks_for_rule_id(rule['id'])
    replica_rse_ids = set(lock['rse_id'] for lock in locks)
    return len(replica_rse_ids) == 1 and list(replica_rse_ids)[0] == rse['id']
```

Which is exactly what I would expect: check whether all the replicas locked by this rule exist on the RSE we’re trying to decommission. ‘Replicas’, without any qualifier before it, means ‘file replicas’. But then:

* `_has_only_one_replica_on_rse`
* `"""Check if the only replica is on the RSE."""`
* `# No replica is locked to begin with -> treat as having the last replica on this RSE`

Reply:
Ah! I now understand what you meant by a better function name. How about _has_all_replicas_on_rse?